File: range-sum-of-bst.py
"""
https://leetcode.com/problems/range-sum-of-bst/submissions/
"""


# This just traverses the whole tree -- we don't do anything special knowing
# that it's a binary search tree.
class TraverseWholeTree:
    def rangeSumBST(self, root, low: int, high: int) -> int:
        def dfs(node):

            if node:
                if low <= node.val <= high:
                    self.ans += node.val
                dfs(node.left)
                dfs(node.right)

        self.ans = 0
        dfs(root)

        return self.ans


# This one just traverses the parts of the tree that it has to.
class Solution:
    def rangeSumBST(self, root, low: int, high: int) -> int:
        def dfs(node):

            if node:
                if low <= node.val <= high:
                    self.ans += node.val

                if node.val > low:
                    dfs(node.left)

                if node.val < high:
                    dfs(node.right)

        self.ans = 0
        dfs(root)
        return self.ans

File: range-sum-of-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} low
 * @param {number} high
 * @return {number}
 */
var rangeSumBSTWholeTree = function(root, low, high) {
    var dfs = function(node) {

        if (node){
            if (low <= node.val & node.val <= high){
                ans += node.val;
            };
            dfs(node.left);
            dfs(node.right);

        };

    };

    var ans = 0;
    dfs(root)
    return ans
};

var rangeSumBST = function(root, low, high) {
    var dfs = function(node) {

        if (node){
            if (low <= node.val & node.val <= high){
                ans += node.val;
            };

            if (node.val > low){
                dfs(node.left);
            };

            if (node.val < high){
                dfs(node.right);
            };

        };

    };

    var ans = 0;
    dfs(root)
    return ans
};
